docs/plugins/hook.rst

```diff
@@ -0,0 +1,77 @@
+Hook Plugin
+===============
+
+Internally, beets sends events to plugins when an action finishes. These can
+range from importing a song (``import``) to beets exiting (``cli_exit``), and
+provide a very flexible way to perform actions based on the events. This plugin
+allows you to run commands when an event is emitted by beets, such as syncing
+your library with another drive when the library is updated.
+
+Hooks are currently run in the order defined in the configuration, however this
+is dependent on beets itself (may change in the future) and cannot be controlled
+by this plugin.
+
+.. _hook-configuration:
+
+Configuration
+-------------
+
+To configure the plugin, make a ``hook:`` section in your configuration
+file. The available options are:
+
+- **hooks**: A list of events and the commands to run (see :ref:`individual-hook-configuration`)
+  Default: Empty.
+- **substitute_event**: The string to replace in each command with the name of
+  the event executing it. This can be used to allow one script to act
+  differently depending on the event it was called by. Can be individually
+  overridden (see :ref:`individual-hook-configuration`).
+  Default: ``%EVENT%``
+- **shell**: Run each command in a shell. Can be individually overridden (see
+  :ref:`individual-hook-configuration`).
+  Default: ``yes``
```

Comment:
So after implementing `Popen` with `shlex`, it seems to work as we want. The issue now is the error message that is outputted by `Popen` when the command doesn't exist isn't exactly "user friendly":

```
Traceback (most recent call last):
  File "./beet", line 23, in <module>
    beets.ui.main()
  File "/Users/jack/Documents/Development/Personal/beets/beets/ui/__init__.py", line 1238, in main
    _raw_main(args)
  File "/Users/jack/Documents/Development/Personal/beets/beets/ui/__init__.py", line 1228, in _raw_main
    subcommand.func(lib, suboptions, subargs)
  File "/Users/jack/Documents/Development/Personal/beets/beets/ui/commands.py", line 965, in import_func
    import_files(lib, paths, query)
  File "/Users/jack/Documents/Development/Personal/beets/beets/ui/commands.py", line 942, in import_files
    session.run()
  File "/Users/jack/Documents/Development/Personal/beets/beets/importer.py", line 320, in run
    pl.run_parallel(QUEUE_SIZE)
  File "/Users/jack/Documents/Development/Personal/beets/beets/util/pipeline.py", line 347, in run
    self.coro.send(msg)
  File "/Users/jack/Documents/Development/Personal/beets/beets/util/pipeline.py", line 160, in coro
    task = func(*(args + (task,)))
  File "/Users/jack/Documents/Development/Personal/beets/beets/importer.py", line 1400, in manipulate_files
    task.finalize(session)
  File "/Users/jack/Documents/Development/Personal/beets/beets/importer.py", line 543, in finalize
    self._emit_imported(session.lib)
  File "/Users/jack/Documents/Development/Personal/beets/beets/importer.py", line 569, in _emit_imported
    plugins.send('album_imported', lib=lib, album=self.album)
  File "/Users/jack/Documents/Development/Personal/beets/beets/plugins.py", line 458, in send
    result = handler(**arguments)
  File "/Users/jack/Documents/Development/Personal/beets/beets/plugins.py", line 123, in wrapper
    return func(*args, **kwargs)
  File "/Users/jack/Documents/Development/Personal/beets/beetsplug/hook.py", line 54, in hook_function
    subprocess.Popen(command_pieces).wait()
  File "/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 710, in __init__
    errread, errwrite)
  File "/usr/local/Cellar/python/2.7.11/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 1335, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory
```

What do you think of catching the `OSError` and re-throwing it with a better message (i.e. "OSError: [Errno 2] No such file or directory: my_command"). We could also preserve the stack trace, although it doesn't exactly contain any useful information.